src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,1055 @@
+        }
+        self.records.iter().max_by_key(|(k, _)| k).unwrap().0
+    }
+
+    pub fn get_percentile_90(&mut self) -> u64 {
+        if self.records.is_empty() {
+            return 0;
+        }
+        let mut v = self.records.make_contiguous().to_vec();
+        v.sort_by_key(|k| k.0);
+        v[((self.records.len() - 1) as f64 * 0.90) as usize].0
+    }
+
+    pub fn slope(&self) -> f64 {
+        if self.records.len() <= 1 {
+            return 0.0;
+        }
+
+        // calculate the average of left and right parts
+        let half = self.records.len() / 2;
+        let mut left = 0;
+        let mut right = 0;
+        for (i, r) in self.records.iter().enumerate() {
+            if i < half {
+                left += r.0;
+            } else if self.records.len() - i - 1 < half {
+                right += r.0;
+            }
+        }
+        // use the two averages with the time span of oldest and latest records
+        // to get a slope
+        let elapsed = duration_to_sec(
+            self.records
+                .back()
+                .unwrap()
+                .1
+                .duration_since(self.records.front().unwrap().1),
+        );
+        (right - left) as f64 / half as f64 / (elapsed / 2.0)
+    }
+
+    pub fn trend(&self) -> Trend {
+        if self.records.len() <= 1 {
+            return Trend::NoTrend;
+        }
+
+        // calculate the average of left and right parts
+        let half = self.records.len() / 2;
+        let mut left = 0;
+        let mut right = 0;
+        for (i, r) in self.records.iter().enumerate() {
+            if i < half {
+                left += r.0;
+            } else if self.records.len() - i - 1 < half {
+                right += r.0;
+            }
+        }
+
+        // decide if there is a trend by the two averages
+        // adding 2 here is to give a tolerance
+        if right > left + 2 {
+            Trend::Increasing
+        } else if right < left - 2 {
+            Trend::Decreasing
+        } else {
+            Trend::NoTrend
+        }
+    }
+}
+
+// CFFlowChecker records some statistics and states related to one CF.
+// These statistics falls into five categories:
+//   * memtable
+//   * L0 files
+//   * flush flow
+//   * L0 flow (compaction read flow of L0)
+//   * pending compaction bytes
+// And all of them are collected from the hook of RocksDB's event listener.
+struct CFFlowChecker {
+    // Memtable related
+    last_num_memtables: Smoother<20>,
+    memtable_debt: f64,
+    init_speed: bool,
+
+    // L0 files related
+    // last number of l0 files right after flush or L0 compaction
+    last_num_l0_files: u64,
+    // last number of l0 files right after flush
+    last_num_l0_files_from_flush: u64,
+    // a few records of number of l0 files right after L0 compaction
+    // As we know, after flush the number of L0 files must increase by 1,
+    // whereas, after L0 compaction the number of L0 files must decrease a lot
```

Comment:
Memtable has the same pattern, why not use similar method?

Reply:
the number of memtables is steady, didn't see any case that the number of memtable will change up and down dramatically.

Reply:
Check this, the frequency is higher, but it's still zigzagging. ![{DD13BF97-3DB5-4255-B842-DE6FEA6CBC8D}](https://user-images.githubusercontent.com/27005812/128491234-34b8a753-145c-4eb1-a27e-e37afeb3fd62.png)
Anyway, no objection here, just curious.